README.txt
= RVideo

== DESCRIPTION:

RVideo allows you to inspect and process video files.

== INSTALL:

Installation is a little involved. First, install the gem:

  sudo gem install rvideo
  
Next, install ffmpeg and (possibly) other related libraries. This is
documented elsewhere on the web, and can be a headache. If you are on OS X,
the Darwinports build is reasonably good (though not perfect). Install with:

  sudo port install ffmpeg

Or, for a better build (recommended), add additional video- and audio-related
libraries, like this:

  sudo port install ffmpeg +lame +libogg +vorbis +faac +faad +xvid +x264 +a52
  
Most package management systems include a build of ffmpeg, but many include a
poor build. So you may need to compile from scratch.

If you want to create Flash Video files, also install flvtool2:

  sudo gem install flvtool2

Once ffmpeg and RVideo are installed, you're set. 

== SYNOPSIS:

To inspect a file, initialize an RVideo file inspector object. See the 
documentation for details.

A few examples:

  file = RVideo::Inspector.new(:file => "#{APP_ROOT}/files/input.mp4")
  
  file = RVideo::Inspector.new(:raw_response => @existing_response)
  
  file = RVideo::Inspector.new(:file => "#{APP_ROOT}/files/input.mp4",
                                :ffmpeg_binary => "#{APP_ROOT}/bin/ffmpeg")

  file.fps        # "29.97"
  file.duration   # "00:05:23.4"

To transcode a video, initialize a Transcoder object.

  transcoder = RVideo::Transcoder.new

Then pass a command and valid options to the execute method

  recipe = "ffmpeg -i $input_file$ -ar 22050 -ab 64 -f flv -r 29.97 -s"
  recipe += " $resolution$ -y $output_file$"
  recipe += "\nflvtool2 -U $output_file$"
  begin
    transcoder.execute(recipe, {:input_file => "/path/to/input.mp4",
      :output_file => "/path/to/output.flv", :resolution => "640x360"})
  rescue TranscoderError => e
    puts "Unable to transcode file: #{e.class} - #{e.message}"
  end

If the job succeeds, you can access the metadata of the input and output
files with:

  transcoder.original     # RVideo::Inspector object
  transcoder.processed    # RVideo::Inspector object

If the transcoding succeeds, the file may still have problems. RVideo
will populate an errors array if the duration of the processed video
differs from the duration of the original video, or if the processed
file is unreadable.

== FEATURES/PROBLEMS:

== REQUIREMENTS:

Thanks to Peter Boling for early work on RVideo.

Contribute to RVideo! If you want to help out, there are a few things you can 
do.

- Use, test, and submit bugs/patches
- We need a RVideo::Tools::Mencoder class to add mencoder support.
- Other tool classes would be great - On2, mp4box, Quicktime (?), etc.
- Submit other fixes, features, optimizations, and refactorings

If RVideo is useful to you, you may also be interested in RMovie, another Ruby
video library. See http://rmovie.rubyforge.org/ for more.

Finally, watch for Zencoder, a commercial video transcoder built by Slantwise 
Design. Zencoder uses RVideo for its video processing, but adds file queuing,  
distributed transcoding, a web-based transcoder dashboard, and more. See 
http://zencoder.tv or http://slantwisedesign.com for more.

Copyright (c) 2007 Jonathan Dahl and Slantwise Design. Released under the MIT 
license.

History.txt
== 0.9.5 2008-09-09
Note:
* Moving hosting to GitHub.
  http://github.com/zencoder/rvideo/tree

2 major enhancements:
* Large files, or some tools, would dump out so much output that it could run a server out of memory.
  Changed the way that output is parsed.  Instead of pulling it all into memory, the log file is sent to 
  disk, and then post-processed.
* Added support for yamdi (http://yamdi.sourceforge.net/) for FLV metadata injection.  This is much faster 
  than flvtool2 on huge files, and uses less memory as well.


== 0.9.4 2007-12-12

3 major enhancements:
* Changed transcoder interface. The preferred interface is now to pass the input file to RVideo::Transcoder.new. This allows you to create multiple output files from the same Transcoder object, and inspects the input file metadata before the job is transcoded.
* Added screengrab functionality to the Inspector class. Fire up an inspector instance on a file, and you can take one or more screenshots through inspected_file.capture_frame("50%").
* Added resolution/aspect support. Pass :width and :height parameters to a Transcoder instance, and include $resolution$ or $aspect_ratio$ in the recipe. 

4 minor enhancements:
* Remove old/unnecessary files and features
* Three additional ffmpeg results: unsupported codec and no output streams
* One additional flvtool2 result: empty input file
* Check that input file exists earlier in the transcoding process

== 0.9.3 2007-10-30

One minor enhancement:
* Reraise all unhandled RVideo exceptions as TranscoderErrors

== 0.9.2 2007-10-30

One minor bug fix:
* Correctly parse invalid files, where duration and bitrate are N/A, but start: 0.000 is included

== 0.9.1 2007-10-11

One major enhancement:
* Added Mencoder support. (Andre Medeiros)

Two minor enhancements: 
* Added total_time method to RVideo::Transcoder instances.
* Added another error condition for ffmpeg - codec not found.

Two notes:
* Tried and tested using open3 and open4 for command execution, but mencoder would unexpectedly hang with certain files when using these. Reverted these changes.
* Mencoder has basic unit tests, but needs more tests. In particular, example output should be added for a variety of cases (especially failures and errors).

== 0.9.0 2007-09-27

* Public RVideo release.

== 0.8.0 2007-09-27

* RVideo rewrite.


lib/rvideo/transcoder.rb
module RVideo # :nodoc:
  class Transcoder
    
    attr_reader :executed_commands, :processed, :errors, :warnings, :total_time
    
    #
    # To transcode a video, initialize a Transcoder object:
    #
    #   transcoder = RVideo::Transcoder.new("/path/to/input.mov")
    #
    # Then pass a recipe and valid options to the execute method
    #
    #   recipe = "ffmpeg -i $input_file$ -ar 22050 -ab 64 -f flv -r 29.97 -s"
    #   recipe += " $resolution$ -y $output_file$"
    #   recipe += "\nflvtool2 -U $output_file$"
    #   begin
    #     transcoder.execute(recipe, {:output_file => "/path/to/output.flv", 
    #       :resolution => "640x360"})
    #   rescue TranscoderError => e
    #     puts "Unable to transcode file: #{e.class} - #{e.message}"
    #   end
    #
    # If the job succeeds, you can access the metadata of the input and output
    # files with:
    #
    #   transcoder.original     # RVideo::Inspector object
    #   transcoder.processed    # RVideo::Inspector object
    #
    # If the transcoding succeeds, the file may still have problems. RVideo
    # will populate an errors array if the duration of the processed video
    # differs from the duration of the original video, or if the processed
    # file is unreadable.
    # 
    
    def initialize(input_file = nil)
      # Allow a nil input_file for backwards compatibility. (Change at 1.0?)
      check_input_file(input_file)
      
      @input_file = input_file
      @executed_commands = []
      @errors = []
      @warnings = []
    end
    
    def original
      @original ||= Inspector.new(:file => @input_file)
    end
    
    #
    # Configure logging. Pass a valid Ruby logger object.
    #
    #   logger = Logger.new(STDOUT)
    #   RVideo::Transcoder.logger = logger
    #
    
    def self.logger=(l)
      @logger = l
    end
    
    def self.logger
      if @logger.nil?
        @logger = Logger.new('/dev/null')
      end
      
      @logger
    end
    
    #
    # Requires a command and a hash of various interpolated options. The
    # command should be one or more lines of transcoder tool commands (e.g.
    # ffmpeg, flvtool2). Interpolate options by adding $option_key$ to the
    # recipe, and passing :option_key => "value" in the options hash.
    #
    #   recipe = "ffmpeg -i $input_file$ -ar 22050 -ab 64 -f flv -r 29.97 
    #   recipe += "-s $resolution$ -y $output_file$"
    #   recipe += "\nflvtool2 -U $output_file$"
    #
    #   transcoder = RVideo::Transcoder.new("/path/to/input.mov")
    #   begin
    #     transcoder.execute(recipe, {:output_file => "/path/to/output.flv", :resolution => "320x240"})
    #   rescue TranscoderError => e
    #     puts "Unable to transcode file: #{e.class} - #{e.message}"
    #   end
    #
    
    def execute(task, options = {})
      t1 = Time.now
      
      if @input_file.nil?
        @input_file = options[:input_file]
      end
      
      Transcoder.logger.info("\nNew transcoder job\n================\nTask: #{task}\nOptions: #{options.inspect}")
      parse_and_execute(task, options)      
      @processed = Inspector.new(:file => options[:output_file])
      result = check_integrity
      Transcoder.logger.info("\nFinished task. Total errors: #{@errors.size}\n")
      @total_time = Time.now - t1
      result
    rescue TranscoderError => e
      raise e
    rescue Exception => e
      Transcoder.logger.error("[ERROR] Unhandled RVideo exception: #{e.class} - #{e.message}\n#{e.backtrace}")
      raise TranscoderError::UnknownError, "Unexpected RVideo error: #{e.message} (#{e.class})"
    end
        
    private
    
    def check_input_file(input_file)
      if input_file and !FileTest.exist?(input_file.gsub("\"",""))
        raise TranscoderError::InputFileNotFound, "File not found (#{input_file})"
      end
    end
    
    def check_integrity
      precision = 1.1
      if @processed.invalid?
        @errors << "Output file invalid"
      elsif (@processed.duration >= (original.duration * precision) or @processed.duration <= (original.duration / precision))
        @errors << "Original file has a duration of #{original.duration}, but processed file has a duration of #{@processed.duration}"
      end
      return @errors.size == 0
    end
    
    def parse_and_execute(task, options = {})
      raise TranscoderError::ParameterError, "Expected a recipe class (as a string), but got a #{task.class.to_s} (#{task})" unless task.is_a? String
      options = options.merge(:input_file => @input_file)
      
      commands = task.split("\n").compact
      commands.each do |c|
        tool = Tools::AbstractTool.assign(c, options)
        tool.original = @original
        tool.execute
        executed_commands << tool
      end
    end
  end
end
